src/main/java/org/cryptomator/cryptofs/fh/ChunkIO.java

```diff
@@ -16,6 +15,8 @@ class ChunkIO {
 	private final Set<FileChannel> readableChannels = ConcurrentHashMap.newKeySet();
 	private final Set<FileChannel> writableChannels = ConcurrentHashMap.newKeySet();
 
+	private final PriorityMutex synchronizer = new PriorityMutex();
```

Comment:
A conventional ReadWriteLock should suffice:

* readlock: allow multiple threads to enter `write()`
* writelock: only allow a thread to enter `unregisterChannel()` when nobody is `write()`ing

Reply:
which has the advantage that there is no need to write plenty of tests covering concurrency issues with a custom synchronization means.